cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package partition
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	diskFromMountPoint = diskFromMountPointImplWrapper
+	mockedMountPoints  = make(map[string]*mockDisk)
+	isSnapdTest        = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test")
+
+	luksUUIDPatternRe = regexp.MustCompile(`(?m)CRYPT-LUKS2-([0-9a-f]{32})`)
+
+	totalMockedDisks = 0
+)
+
+// Options is a set of options used when querying information about
+// partition and disk devices.
+type Options struct {
+	IsDecryptedDevice bool
+}
+
+// Disk is a single physical disk device that contains partitions.
+type Disk interface {
+	// FindMatchingPartitionUUID finds the partition uuid for a partition matching
+	// the specified label on the disk. Note that for non-ascii labels like
+	// "Some label", the label should be encoded using \x<hex> for potentially
+	// non-safe characters like in "Some\x20Label".
+	FindMatchingPartitionUUID(string) (string, error)
+
+	// MountPointIsFromDisk returns whether the specified mountpoint corresponds
+	// to a partition on the disk.
+	MountPointIsFromDisk(string, *Options) (bool, error)
+
+	// String returns a representation of the disk that may not be unique, it
+	// is meant for testing/debugging.
+	String() string
```

Comment:
yes actually I added this just for tests, the alternative is some weird export_test.go code which I'm happy to resurrect. it's weird because it the exported methods all return a `Disk` interface, so to add something there I need to type cast it or I need to invent a new interface in export_test.go which adds String() and embeds the normal Disk interface and then also create a different function which returns the test interface

Reply:
I see, do we use it for %s ? or just for tests?

Reply:
just for tests

Reply:
ok, probably best to be explicit then and do something like:

// Dev returns the string "major:minor" for the disk device for identification, it is not unique but useful for tests.